File: dashboard/src/lib/audit-chain.ts
// HMAC chain helpers — must produce IDENTICAL bytes to the Python
// implementation in ats/storage/audit_chain.py for any given record.
//
// Canonical JSON contract:
//   - Object keys sorted lexicographically (UTF-16 code-unit order;
//     same as Python's sort_keys for ASCII keys).
//   - No whitespace. Numbers / strings / null / booleans serialised
//     with the standard JS JSON rules. Arrays preserve order.
//   - UTF-8 encoded on the wire.
// Hash input layout:
//   HMAC-SHA256(secret, prev_hash || canonical_json(record_view))

import { createHmac } from "crypto";

export const ZERO_HASH = Buffer.alloc(32);

let CACHED_KEY: Buffer | null = null;

function resolveKey(): Buffer {
  if (CACHED_KEY) return CACHED_KEY;
  const raw = (process.env.ATS_AUDIT_HMAC_KEY ?? "").trim();
  if (raw) {
    try {
      CACHED_KEY = Buffer.from(raw, "hex");
    } catch (e) {
      throw new Error(`ATS_AUDIT_HMAC_KEY must be hex-encoded: ${e}`);
    }
  } else {
    // Dev fallback. Production should set the env var.
    CACHED_KEY = Buffer.alloc(32);
  }
  return CACHED_KEY;
}

export function resetKeyCache(): void {
  CACHED_KEY = null;
}

export function nowIso(): string {
  // toISOString already produces millisecond precision + 'Z' suffix —
  // matches what Python emits via isoformat(timespec='milliseconds').
  return new Date().toISOString();
}

export function canonicalJson(value: unknown): string {
  if (value === null) return "null";
  if (typeof value === "number" || typeof value === "boolean") {
    return JSON.stringify(value);
  }
  if (typeof value === "string") return JSON.stringify(value);
  if (Array.isArray(value)) {
    return "[" + value.map(canonicalJson).join(",") + "]";
  }
  if (typeof value === "object") {
    const keys = Object.keys(value as Record<string, unknown>).sort();
    return (
      "{" +
      keys
        .map(
          (k) =>
            JSON.stringify(k) +
            ":" +
            canonicalJson((value as Record<string, unknown>)[k]),
        )
        .join(",") +
      "}"
    );
  }
  // undefined or unknown — coerce to null to match Python behaviour
  // for json.dumps which would raise; we choose stable null instead.
  return "null";
}

export interface AuditRecordView {
  actor_kind: string;
  actor_user_id: number | null;
  created_at: string;
  kind: string;
  org_id: number;
  payload: Record<string, unknown>;
  target_id: number | null;
  target_kind: string | null;
}

export function recordView(args: {
  org_id: number;
  actor_user_id: number | null;
  actor_kind: string;
  kind: string;
  target_kind: string | null;
  target_id: number | null;
  payload: Record<string, unknown>;
  created_at: string;
}): AuditRecordView {
  // Field order here is informational; canonicalJson sorts keys.
  return {
    actor_kind: args.actor_kind,
    actor_user_id: args.actor_user_id,
    created_at: args.created_at,
    kind: args.kind,
    org_id: args.org_id,
    payload: args.payload,
    target_id: args.target_id,
    target_kind: args.target_kind,
  };
}

export function computeHash(
  prevHash: Buffer,
  view: AuditRecordView,
): Buffer {
  const h = createHmac("sha256", resolveKey());
  h.update(prevHash);
  h.update(canonicalJson(view), "utf8");
  return h.digest();
}

File: ats/storage/audit_chain.py
"""HMAC chain helpers for audit_log (sub-project #5b).

Canonical JSON contract — both Python and the TypeScript dashboard MUST
produce the SAME bytes for a given record, or hashes won't match across
languages. The contract:

- Object keys are sorted lexicographically (UTF-16 code-unit order, the
  default of both ``sorted()`` in Python and ``Array.prototype.sort()``
  in JavaScript for ASCII keys, which is all we use).
- No whitespace anywhere.
- Strings escaped per JSON spec, encoded as UTF-8 on the wire.
- Numbers serialized normally (integers without trailing ``.0``).
- ``null`` for nulls.
- Arrays preserve order.

Hash input layout:

    HMAC-SHA256(secret, prev_hash || canonical_json(record_view))

where ``record_view`` is the dict ``{actor_kind, actor_user_id, created_at,
kind, org_id, payload, target_id, target_kind}`` (id and prev_hash/hash
intentionally excluded — id is opaque, prev_hash is in the input bytes
already).
"""

from __future__ import annotations

import hashlib
import hmac
import json
import os
from datetime import datetime, timezone
from typing import Any

ZERO_HASH = bytes(32)

# Cache the resolved key bytes between calls. Resolved lazily so unit
# tests can monkeypatch the env without restarting.
_KEY: bytes | None = None


def _resolve_key() -> bytes:
    """Read ATS_AUDIT_HMAC_KEY (hex). Falls back to a dev-only constant
    so demo deployments work without configuration; production should
    override via env."""
    global _KEY
    if _KEY is not None:
        return _KEY
    raw = os.environ.get("ATS_AUDIT_HMAC_KEY", "").strip()
    if raw:
        try:
            _KEY = bytes.fromhex(raw)
        except ValueError as exc:  # pragma: no cover
            raise RuntimeError(
                f"ATS_AUDIT_HMAC_KEY must be hex-encoded: {exc}"
            ) from exc
    else:
        # Dev fallback. CHANGE THIS in production via env.
        _KEY = b"\x00" * 32
    return _KEY


def reset_key_cache() -> None:
    """Tests use this after monkey-patching the env."""
    global _KEY
    _KEY = None


def now_iso() -> str:
    """ISO-8601 with millisecond precision, UTC, trailing 'Z'.

    Both Python and JS render this exact format (``new Date().toISOString()``
    matches), so the same datetime hashes identically across languages.
    """
    return (
        datetime.now(timezone.utc)
        .isoformat(timespec="milliseconds")
        .replace("+00:00", "Z")
    )


def canonical_json(value: Any) -> bytes:
    """Sorted-keys, no-whitespace JSON in UTF-8 — matches the TS helper.

    Uses ``json.dumps`` with sort_keys=True; Python's default sort of
    string keys is lexicographic over Unicode code points, matching JS's
    default ``Array.sort()`` for ASCII. Our keys are all ASCII.
    """
    return json.dumps(
        value, sort_keys=True, separators=(",", ":"), ensure_ascii=False
    ).encode("utf-8")


def record_view(
    *,
    org_id: int,
    actor_user_id: int | None,
    actor_kind: str,
    kind: str,
    target_kind: str | None,
    target_id: int | None,
    payload: dict[str, Any],
    created_at: str,
) -> dict[str, Any]:
    """Build the dict that gets hashed. Stable shape regardless of writer.

    Keys are alphabetised by ``canonical_json`` so the order here is
    informational; we still pass an ordered dict to make the reading
    clearer.
    """
    return {
        "actor_kind": actor_kind,
        "actor_user_id": actor_user_id,
        "created_at": created_at,
        "kind": kind,
        "org_id": org_id,
        "payload": payload,
        "target_id": target_id,
        "target_kind": target_kind,
    }


def compute_hash(prev_hash: bytes, view: dict[str, Any]) -> bytes:
    return hmac.new(
        _resolve_key(),
        prev_hash + canonical_json(view),
        hashlib.sha256,
    ).digest()
